Userland/Libraries/LibJS/Runtime/DeclarativeEnvironment.h

```diff
@@ -65,20 +69,50 @@ class DeclarativeEnvironment : public Environment {
 
     virtual void visit_edges(Visitor&) override;
 
-private:
-    virtual bool is_declarative_environment() const override { return true; }
+    class BindingAndIndex {
+    public:
+        Binding& binding()
+        {
+            if (m_referenced_binding)
+                return *m_referenced_binding;
+            return m_temporary_binding;
+        }
+
+        BindingAndIndex(Binding* binding, Optional<size_t> index_)
+            : index(move(index_))
+            , m_referenced_binding(binding)
+        {
+        }
+
+        explicit BindingAndIndex(Binding temporary_binding)
+            : m_temporary_binding(move(temporary_binding))
+        {
+        }
+
+        Optional<size_t> index;
+
+    private:
+        Binding* m_referenced_binding { nullptr };
+        Binding m_temporary_binding {};
+    };
+
+    friend class ModuleEnvironment;
 
-    Optional<size_t> find_binding_index(FlyString const& name) const
+    virtual Optional<BindingAndIndex> find_binding_and_index(FlyString const& name) const
```

Comment:
Since this is a quite fundamental operation perhaps we should try to figure out a way to do this without virtual function calls? I couldn't really think of one except having ModuleEnvironment copy the implementation of DeclarationEnvironment for all functions.

The coordinate caching optimization is still working and I don't have actual measurements so it might very well not be a big deal.

Reply:
Let's just roll with it and see if anything falls over